l3kernel/l3luatex.dtx

```diff
@@ -554,6 +616,61 @@ end
 %    \end{macrocode}
 % \end{macro}
 %
+% \begin{macro}{ltx.utils.require}
+%   Loads a Lua module. This function loads the module just like the standard
+%   Lua global function |require|, with two key differences:
+%   \begin{enumerate}
+%     \item This function is guaranteed not to throw an error.
+%     \item This function does \emph{not} return the loaded module; instead, it
+%           returns |nil| on success and an error message on failure.
+%   \end{enumerate}
+%    \begin{macrocode}
+local function require(name)
+  if package_loaded[name] then
+    return
+  end
+
+  local failure_details = {}
+  for _, searcher in ipairs(package_searchers) do
+    local loader, data = searcher(name)
+    if type(loader) == 'function' then
+      package_loaded[name] = loader(name, data) or true
+      return
+    elseif type(loader) == 'string' then
+      failure_details[#failure_details + 1] = loader
+    end
+  end
+
+  return table_concat(failure_details, '\n')
+end
```

Comment:
I would prefer an interface which can also be used if the return value is needed and which follows the Lua convention of returning something which evaluates as false followed by the error when an error happens.

Reply:
That interface would make much more sense. And it turns out that with that interface, I can just use `pcall` instead of reimplementing `require`, which simplifies things quite a bit.

Reply:
Personally I preferred the older version since it separates error messages better. By moving `require` completely into a `pcall` we don't just catch errors due to the module not being found but also error from failing to load the module. Of course, if the actual error is that a module failed to load, having a "Lua module `...' not found." error message would be very misleading.

Reply:
Sure. I've (re)added the `try_require` function. 

One problem now though is that if the module throws an error when it's loaded, it puts TeX into a bad state. If you were to do `\directlua{require "something-that-throws-an-error"}`, the Lua interpreter will error out, but this won't propagate to the TeX end. With this change, `\lua_load_module:n {something-that-throws-an-error}` also puts TeX into a bad state since `put_next(c_{true/false}_bool)` is never called. We get a nice Lua stack trace either way though, so I think that this is fine.